File: core/src/main/java/pcb/uwu/amount/composite/mechanics/Speed.java
package pcb.uwu.amount.composite.mechanics;

import pcb.uwu.amount.quantity.Length;
import pcb.uwu.amount.quantity.Time;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.CompositeUnitAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.mechanics.AccelerationUnit;
import pcb.uwu.units.composite.mechanics.SpeedUnit;
import pcb.uwu.units.quantity.LengthUnit;
import pcb.uwu.units.quantity.TimeUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class Speed extends CompositeUnitAmount<SpeedUnit> {

	// region constructors

	public Speed(Number value, SpeedUnit unit) {
		super(value, unit);
	}

	public Speed(Number value, Magnitude magnitude, SpeedUnit unit) {
		super(value, magnitude, unit);
	}

	public Speed(String value, SpeedUnit unit) {
		super(value, unit);
	}

	public Speed(String value, Magnitude magnitude, SpeedUnit unit) {
		super(value, magnitude, unit);
	}

	public Speed(BigDecimal value, SpeedUnit unit) {
		super(value, unit);
	}

	public Speed(BigDecimal value, Magnitude magnitude, SpeedUnit unit) {
		super(value, magnitude, unit);
	}

	public Speed(BigDecimalAmount amount, SpeedUnit unit) {
		super(amount, unit);
	}

	public Speed(BigDecimalAmount amount, Magnitude magnitude, SpeedUnit unit) {
		super(amount, magnitude, unit);
	}

	// endregion

	// region implement UnitAmount

	@Override
	public Speed plus(UnitAmount<SpeedUnit> other, MathContext mathContext) {
		return new Speed(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public Speed minus(UnitAmount<SpeedUnit> other, MathContext mathContext) {
		return new Speed(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public Speed multipliedBy(BigDecimal other, MathContext mathContext) {
		return new Speed(multipliedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Speed dividedBy(BigDecimal other, MathContext mathContext) {
		return new Speed(dividedByScalar(other, mathContext), getUnit());
	}

	@Override
	public Speed convertTo(SpeedUnit unit) {
		return new Speed(getAmountIn(unit), unit);
	}

	// endregion

	// region composition

	public Acceleration dividedBy(Time time, MathContext mathContext) {
		BigDecimalAmount amount = getAmount()
				.dividedBy(time.getAmount().getValue(), mathContext);
		AccelerationUnit unit = new AccelerationUnit(getUnit(), time.getUnit());

		return new Acceleration(amount, unit);
	}

	public Length multipliedBy(Time time, MathContext mathContext) {
		LengthUnit lengthUnit = getUnit().getUnitCounter().findUnit(LengthUnit.class);
		TimeUnit timeUnit = getUnit().getUnitCounter().findUnit(TimeUnit.class);
		BigDecimalAmount amount = getAmount()
				.multipliedBy(time.getAmountIn(timeUnit), mathContext);

		return new Length(amount, lengthUnit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/derived/mechanics/Speed.kt
package pcb.uwu.amount.derived.mechanics

import pcb.uwu.amount.base.Length
import pcb.uwu.amount.base.Time
import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.base.LengthUnit
import pcb.uwu.unit.base.TimeUnit
import pcb.uwu.unit.derived.mechanics.AccelerationUnit
import pcb.uwu.unit.derived.mechanics.PaceUnit
import pcb.uwu.unit.derived.mechanics.SpeedUnit
import pcb.uwu.util.UnitAmountUtils

open class Speed : CompositeUnitAmount<SpeedUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: SpeedUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: SpeedUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(speed: UnitAmount<SpeedUnit>) =
        Speed(amount = this.amount + (speed to this.unit).amount,
              unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(speed: UnitAmount<SpeedUnit>) =
        Speed(amount = this.amount - (speed to this.unit).amount,
              unit = this.unit)

    override fun times(scalar: Number) =
        Speed(amount = this.amount * scalar,
              unit = this.unit)

    override fun div(scalar: Number) =
        Speed(amount = this.amount / scalar,
              unit = this.unit)

    override fun to(unit: SpeedUnit) =
        Speed(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
              unit = unit)

    override fun invert() =
        Pace(amount = this.amount.invert(),
             unit = PaceUnit(this.unit))

    // endregion

    // region composition

    open operator fun div(time: Time) =
        Acceleration(amount = this.amount / time.amount,
                     unit = AccelerationUnit(lengthUnit = this.unit,
                                             timeUnit = time.unit))

    open operator fun div(acceleration: Acceleration) =
        Time(amount = super.div(acceleration).amount,
             unit = this.unit.unitCounter.findUnit(TimeUnit::class)!!)

    open operator fun times(time: Time) =
        Length(amount = super.times(time).amount,
               unit = this.unit.unitCounter.findUnit(LengthUnit::class)!!)

    // endregion
}
